src/main/java/net/starlark/java/eval/StarlarkInt.java

```diff
@@ -505,10 +505,32 @@ public static StarlarkInt subtract(StarlarkInt x, StarlarkInt y) {
 
   /** Returns x * y. */
   public static StarlarkInt multiply(StarlarkInt x, StarlarkInt y) {
-    if (x instanceof Int32 && y instanceof Int32) {
-      long xl = ((Int32) x).v;
-      long yl = ((Int32) y).v;
-      return StarlarkInt.of(xl * yl);
+    try {
+      long xl = x.toLongFast();
+      long yl = y.toLongFast();
+
+      // Copy-paste from Guava LongMath.saturatedMultiply https://git.io/JIlZP
+
+      long zl = xl * yl;
+
+      int leadingZeros =
+        Long.numberOfLeadingZeros(xl)
+          + Long.numberOfLeadingZeros(~xl)
+          + Long.numberOfLeadingZeros(yl)
+          + Long.numberOfLeadingZeros(~yl);
+
+      if (leadingZeros > Long.SIZE + 1) {
+        return StarlarkInt.of(zl);
```

Comment:
Is this really faster than retaining the int32*int32 special case? The sheer number of branches alone makes me skeptical.

Reply:
No, it is slower. I was running the wrong benchmark. Thank for noticing! I'll restore the int32*int32 special case.